### PR TITLE
Fix bug in ShardedTensorMetadata serde.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63902

The 'memory_format' field was not being serialized correctly and used
the same encoding for different fields.

Differential Revision: [D30527324](https://our.internmc.facebook.com/intern/diff/D30527324/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

### CODE REVIEW

File: test/distributed/_sharded_tensor/test_sharded_tensor.py [hunk 1]
```diff
@@ -122,6 +124,52 @@ def wrapper(self):
...
+
+        dtypes = [
+            torch.float, torch.double, torch.cfloat, torch.cdouble, torch.half,
+            torch.bfloat16, torch.uint8, torch.int8, torch.short, torch.int,
+            torch.long, torch.bool]
+
+        layouts = [torch.strided, torch.sparse_coo]
+        requires_grad = [True, False]
+        memory_format = [torch.contiguous_format, torch.channels_last, torch.preserve_format]
+        pin_memory = [True, False]
```
Comment (bowangbj) [inline test/distributed/_sharded_tensor/test_sharded_tensor.py:160]: nit:  require_grads  memory_formats  pin_memories

File: test/distributed/_sharded_tensor/test_sharded_tensor.py [hunk 2]
```diff
@@ -122,6 +124,52 @@ def wrapper(self):
...
+
+        for test in itertools.product(dtypes, layouts, requires_grad, memory_format, pin_memory):
+            dtype, layout, requires_grad, memory_format, pin_memory = test
+
+            expected_st_metadata = _sharded_tensor.ShardedTensorMetadata(
+                shards_metadata, (10, 10), dtype, layout, requires_grad,
```
Comment (bowangbj) [inline test/distributed/_sharded_tensor/test_sharded_tensor.py:166]: maybe use tensor_properties ?

File: test/distributed/_sharded_tensor/test_sharded_tensor.py [hunk 3]
```diff
@@ -122,6 +124,52 @@ def wrapper(self):
...
+            torch.float, torch.double, torch.cfloat, torch.cdouble, torch.half,
+            torch.bfloat16, torch.uint8, torch.int8, torch.short, torch.int,
+            torch.long, torch.bool]
+
+        layouts = [torch.strided, torch.sparse_coo]
+        requires_grad = [True, False]
+        memory_format = [torch.contiguous_format, torch.channels_last, torch.preserve_format]
+        pin_memory = [True, False]
+
+        for test in itertools.product(dtypes, layouts, requires_grad, memory_format, pin_memory):
```
Comment (bowangbj) [inline test/distributed/_sharded_tensor/test_sharded_tensor.py:162]: nit: 
s/test/tensor_properties_input

File: test/distributed/_sharded_tensor/test_sharded_tensor.py [hunk 4]
```diff
@@ -122,6 +124,52 @@ def wrapper(self):
         self.destroy_comms()
     return wrapper
 
+class TestShardedTensorMetadata(TestCase):
+    def test_serde(self):
+        shards_metadata = [
+            ShardMetadata(
```
Comment (bowangbj) [inline test/distributed/_sharded_tensor/test_sharded_tensor.py:130]: nit: shard_metadatas ?

File: test/distributed/_sharded_tensor/test_sharded_tensor.py [hunk 5]
```diff
@@ -122,6 +124,52 @@ def wrapper(self):
         self.destroy_comms()
     return wrapper
 
+class TestShardedTensorMetadata(TestCase):
+    def test_serde(self):
```
Comment (bowangbj) [inline test/distributed/_sharded_tensor/test_sharded_tensor.py:128]: probably rename it to 

test_serialize_and_deserialize 

for better readability ?

**Comment by bowangbj** [comment]
Thanks Pritam for the great fix, just a few small nits.

**Review by bowangbj — COMMENTED** [review]
Thanks Pritam for the great fix, just a few small nits.

File: torch/distributed/_sharded_tensor/api.py [hunk 6]
```diff
@@ -85,7 +85,7 @@ def __getstate__(self):
         elif self.memory_format == torch.channels_last:
             mem_format_encoding = 1
         elif self.memory_format == torch.preserve_format:
-            mem_format_encoding = 1
+            mem_format_encoding = 2
```
Comment (SciPioneer) [inline torch/distributed/_sharded_tensor/api.py:88]: Shall we replace the numeric values here with more readable enum values?

**Comment by pritamdamania87** [comment]
@SciPioneer @bowangbj Can you take another look at this PR? Thanks!

**Review by bowangbj — APPROVED** [review]
Looks great, thanks Pritam!

**Pull request closed without merging**